pilot/pkg/networking/core/v1alpha3/cluster.go

```diff
@@ -202,10 +239,53 @@ func (configgen *ConfigGeneratorImpl) BuildDeltaClusters(proxy *model.Proxy, pus
 		model.XdsLogDetails{AdditionalInfo: fmt.Sprintf("cached:%v/%v", clusterCacheStats.hits, clusterCacheStats.hits+clusterCacheStats.miss)}, true
 }
 
+// getRemovedClusterName finds the name of the removed cluster given a hostname and list of cluster names.
+// subsets is ignored if it is nil
+func getRemovedClusterNames(resNames []string, subsets []string, target string) []string {
+	removed := make([]string, 0)
+	for _, n := range resNames {
+		_, sub, svcHost, _ := model.ParseSubsetKey(n)
+		if svcHost == host.Name(target) && (len(subsets) == 0 || contains(subsets, sub)) {
+			removed = append(removed, n)
+		}
+	}
+	return removed
+}
+
+// getDeletedSubsets detects deleted subsets and returns those names
+func getDeletedSubsets(orig []*networking.Subset, new []*networking.Subset) []string {
+	subsetNames := make([]string, 0)
+	// deleted if exists in orig but not in new
+	ne := make(map[string]struct{})
+	for _, v := range new {
+		ne[v.Name] = struct{}{}
+	}
+	for _, v := range orig {
+		if _, found := ne[v.Name]; !found {
+			subsetNames = append(subsetNames, v.Name)
+		}
+	}
+	return subsetNames
+}
+
 func shouldUseDelta(updates *model.PushRequest) bool {
 	// Use delta when "only" services are modified. We will relax this restriction
 	// as we enhance delta code to handle other complex/mixed config changes.
-	return updates != nil && allConfigKeysOfType(updates.ConfigsUpdated, gvk.ServiceEntry) && len(updates.ConfigsUpdated) > 0
+	return updates != nil && AllConfigKeysAreOfTypes(updates.ConfigsUpdated, []config.GroupVersionKind{gvk.ServiceEntry, gvk.DestinationRule}) && len(updates.ConfigsUpdated) > 0
+}
+
+// AllConfigKeysAreOfTypes returns whether or not the configs provided only consist of the gvks in target.
+func AllConfigKeysAreOfTypes(cfgs map[model.ConfigKey]struct{}, target []config.GroupVersionKind) bool {
+	deduped := make(map[config.GroupVersionKind]struct{}, 0)
+
+	for k := range cfgs {
```

Comment:
nit: You can return if you find a cfg that is not part of target. It would simplify things if target is a `map[gvk]struct{}`